org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/contentassist/JavaDocCompletionProposal.java

```diff
@@ -0,0 +1,235 @@
+/*******************************************************************************
+ * Copyright (c) 2016-2017 Red Hat Inc. and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Based on org.eclipse.jdt.internal.ui.text.javadoc.JavaDocAutoIndentStrategy
+ *
+ * Contributors:
+ *     Red Hat Inc. - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.jdt.ls.core.internal.contentassist;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.jdt.core.ICompilationUnit;
+import org.eclipse.jdt.core.IJavaElement;
+import org.eclipse.jdt.core.IJavaProject;
+import org.eclipse.jdt.core.IMember;
+import org.eclipse.jdt.core.IMethod;
+import org.eclipse.jdt.core.ISourceRange;
+import org.eclipse.jdt.core.IType;
+import org.eclipse.jdt.core.JavaModelException;
+import org.eclipse.jdt.internal.corext.util.MethodOverrideTester;
+import org.eclipse.jdt.internal.corext.util.SuperTypeHierarchyCache;
+import org.eclipse.jdt.ls.core.internal.JDTUtils;
+import org.eclipse.jdt.ls.core.internal.corext.codemanipulation.CodeGeneration;
+import org.eclipse.jdt.ls.core.internal.corext.codemanipulation.StubUtility;
+import org.eclipse.jdt.ls.core.internal.handlers.CompletionResolveHandler;
+import org.eclipse.jdt.ls.core.internal.handlers.JsonRpcHelpers;
+import org.eclipse.jface.text.BadLocationException;
+import org.eclipse.jface.text.IDocument;
+import org.eclipse.jface.text.IRegion;
+import org.eclipse.jface.text.Region;
+import org.eclipse.jface.text.TextUtilities;
+import org.eclipse.lsp4j.CompletionItem;
+import org.eclipse.lsp4j.CompletionItemKind;
+import org.eclipse.lsp4j.InsertTextFormat;
+
+public class JavaDocCompletionProposal {
```

Comment:
How hard is it to refactor JavaDocAutoIndentStrategy in jdt.ui down to jdt.core?

Reply:
There are a few eclipse.jface.text interfaces that would need to move to eclipse.text.  There is a reference to StubUtility and CodeGeneration which is still in the works by Roland.  There are two methods (isSmartMode() and getCompilationUnit()) that drag in all the eclipse.ui stuff such as PlatformUI and IEditorPart.  The class could be split into core and non-core and the smartMode setting plus ICompilationUnit could be set for core so as to avoid eclipse.ui.  There are also a few Preferences to reference/set like we have done before (e.g. Member sort order).

Reply:
There are a lot of changes due to which, I think, we won't be able to use the same code in jdt.ls and jdt.ui

Reply:
@jjohnstn thanks for looking into it. Given the amount of changes and @snjeza's feedback, I'm fine keeping a customized version of that JavaDocAutoIndentStrategy in jdt.ls.